Xamarin.Forms.Core/Picker.cs

```diff
@@ -8,63 +9,325 @@ namespace Xamarin.Forms
 
-		public static readonly BindableProperty TitleProperty = BindableProperty.Create(nameof(Title), typeof(string), typeof(Picker), default(string));
+		public static readonly BindableProperty TitleProperty =
+			BindableProperty.Create(nameof(Title), typeof(string), typeof(Picker), default(string));
 
-		public static readonly BindableProperty SelectedIndexProperty = BindableProperty.Create(nameof(SelectedIndex), typeof(int), typeof(Picker), -1, BindingMode.TwoWay,
-			propertyChanged: (bindable, oldvalue, newvalue) =>
-			{
-				EventHandler eh = ((Picker)bindable).SelectedIndexChanged;
-				if (eh != null)
-					eh(bindable, EventArgs.Empty);
-			}, coerceValue: CoerceSelectedIndex);
+		public static readonly BindableProperty SelectedIndexProperty =
+			BindableProperty.Create(nameof(SelectedIndex), typeof(int), typeof(Picker), -1, BindingMode.TwoWay,
+									propertyChanged: OnSelectedIndexChanged, coerceValue: CoerceSelectedIndex);
+
+		public static readonly BindableProperty ItemsSourceProperty =
+			BindableProperty.Create(nameof(ItemsSource), typeof(IList), typeof(Picker), default(IList),
+									propertyChanged: OnItemsSourceChanged);
+
+		public static readonly BindableProperty SelectedItemProperty =
+			BindableProperty.Create(nameof(SelectedItem), typeof(object), typeof(Picker), null, BindingMode.TwoWay,
+									propertyChanged: OnSelectedItemChanged);
 
 		readonly Lazy<PlatformConfigurationRegistry<Picker>> _platformConfigurationRegistry;
 
 		public Picker()
 		{
-			Items = new ObservableList<string>();
-			((ObservableList<string>)Items).CollectionChanged += OnItemsCollectionChanged;
+			((INotifyCollectionChanged)Items).CollectionChanged += OnItemsCollectionChanged;
 			_platformConfigurationRegistry = new Lazy<PlatformConfigurationRegistry<Picker>>(() => new PlatformConfigurationRegistry<Picker>(this));
 		}
 
-		public IList<string> Items { get; }
+		public IList<string> Items { get; } = new LockableObservableListWrapper();
+
+		public IList ItemsSource
+		{
+			get { return (IList)GetValue(ItemsSourceProperty); }
+			set { SetValue(ItemsSourceProperty, value); }
+		}
 
 		public int SelectedIndex
 		{
 			get { return (int)GetValue(SelectedIndexProperty); }
 			set { SetValue(SelectedIndexProperty, value); }
 		}
 
-		public Color TextColor
+		public object SelectedItem
 		{
+			get { return GetValue(SelectedItemProperty); }
+			set { SetValue(SelectedItemProperty, value); }
+		}
+
+		public Color TextColor {
 			get { return (Color)GetValue(TextColorProperty); }
 			set { SetValue(TextColorProperty, value); }
 		}
 
-		public string Title
-		{
+		public string Title {
 			get { return (string)GetValue(TitleProperty); }
 			set { SetValue(TitleProperty, value); }
 		}
 
+		BindingBase _itemDisplayBinding;
+		public BindingBase ItemDisplayBinding {
+			get { return _itemDisplayBinding; }
+			set {
+				if (_itemDisplayBinding == value)
+					return;
+
+				OnPropertyChanging();
+				var oldValue = value;
```

Comment:
Noticing (long after the fact) oldValue is in fact assigned the newly set value. Won't have any impact the way the code doesn't make use of these values in OnItemDisplayBindingChanged, but may cause future issues.